### ISSUE TITLE
unnecessary warning with autocast 

### ISSUE THREAD

**Issue by maxwellzh** [pytorch/pytorch#67598]
Using autocast like this
```python3
with autocast(enable=(False if device=='cpu' else True)):
    pass
```
When I'm using `device='cpu'`, autocast should be disabled. However, it still raise warning, 
```
.../python3.9/site-packages/torch/autocast_mode.py:141: UserWarning: User provided device_type of 'cuda', but CUDA is not available. Disabling
  warnings.warn('User provided device_type of \'cuda\', but CUDA is not available. Disabling')
```
which I think should be suppressed when `enable=False`.  The code lies at

https://github.com/pytorch/pytorch/blob/97f29bda59deab8c063cf01f0a8ff4321b93c55e/torch/autocast_mode.py#L162

**Comment by glenn-jocher** [comment]
@maxwellzh I'm seeing this also in YOLOv5 when training on CPU-only machines. This is a bug introduced by pytorch with torch 1.10.0.

**Comment by maxwellzh** [comment]
> This is a bug introduced by pytorch with torch 1.10.0

@glenn-jocher Yes, thought this has no influence on the parameter updating, it's quite annoying.

**Comment by mruberry** [comment]
Seems like an easy fix, @mcarilli? 

We would accept a PR fixing an erroneous warning.

**Comment by HugThePenguin** [comment]
I'm not sure whether this is safe or not, but you can just go to 
`.../python3.9/site-packages/torch/autocast_mode.py` 
(the file shown in the message.), go Line141, and comment out the 
`wernings.warn(...) ` 
by adding "#" in front of it.
This will hide the annoying massages.

**Comment by drscotthawley** [comment]
I'm getting this message while using MPS.   Everything is explicitly pushed to torch.device('mps') with nothing set to CUDA anywhere, but still seeing this message.

**Comment by changeling** [comment]
I'm encountering this issue using `mps` on `torch-2.6.0.dev20241210`.
